File: sdks/go/nanolog/handler.go
package nanolog

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"os"
	"runtime"
	"strings"
	"sync"
	"time"
)

type Options struct {
	ServerURL  string
	APIKey     string
	Service    string
	SourceHost string
}

type NanoHandler struct {
	opts       Options
	instanceID string
	queue      chan []byte
	done       chan struct{}
	wg         sync.WaitGroup
	attrs      []slog.Attr
	groups     []string
}

type LogRow struct {
	Timestamp  int64                  `json:"timestamp"`
	Level      string                 `json:"level"`
	Message    string                 `json:"message"`
	Logger     string                 `json:"logger,omitempty"`
	Thread     string                 `json:"thread,omitempty"`
	File       string                 `json:"file,omitempty"`
	Line       int                    `json:"line,omitempty"`
	Service    string                 `json:"service"`
	Host       string                 `json:"host"`
	InstanceID string                 `json:"instance_id"`
	Attributes map[string]interface{} `json:"attributes"`
}

func NewHandler(opts Options) *NanoHandler {
	id, _ := ensureInstanceID()
	h := &NanoHandler{
		opts:       opts,
		instanceID: id,
		queue:      make(chan []byte, 10000),
		done:       make(chan struct{}),
	}

	if h.opts.SourceHost == "" {
		h.opts.SourceHost, _ = os.Hostname()
	}

	// Register asynchronously to not block startup
	go func() {
		if err := registerInstance(opts.ServerURL, opts.APIKey, opts.Service, h.instanceID); err != nil {
			fmt.Fprintf(os.Stderr, "NanoLog Handshake Failed: %v\n", err)
		}
	}()

	h.wg.Add(1)
	go h.runLoop()

	return h
}

func (h *NanoHandler) Enabled(ctx context.Context, level slog.Level) bool {
	return true
}

func (h *NanoHandler) Handle(ctx context.Context, r slog.Record) error {
	row := LogRow{
		Timestamp:  r.Time.UnixNano(),
		Level:      r.Level.String(),
		Message:    r.Message,
		Service:    h.opts.Service,
		Host:       h.opts.SourceHost,
		InstanceID: h.instanceID,
		Attributes: make(map[string]interface{}),
	}

	// Add basic source info
	fs := runtime.CallersFrames([]uintptr{r.PC})
	f, _ := fs.Next()
	row.File = f.File
	row.Line = f.Line
	row.Thread = "goroutine" // Go doesn't expose thread names easily

	// Add attributes
	// 1. Stored attributes (from WithAttrs)
	for _, a := range h.attrs {
		row.Attributes[a.Key] = a.Value.Any()
	}
	// 2. Record attributes
	r.Attrs(func(a slog.Attr) bool {
		row.Attributes[a.Key] = a.Value.Any()
		return true
	})

	data, err := json.Marshal(row)
	if err != nil {
		return err
	}

	select {
	case h.queue <- data:
	default:
		// Drop logging
		fmt.Fprintf(os.Stderr, "NanoLog Queue Full: Dropping log\n")
	}

	return nil
}

func (h *NanoHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	h2 := *h
	h2.attrs = append(h2.attrs, attrs...)
	return &h2
}

func (h *NanoHandler) WithGroup(name string) slog.Handler {
	h2 := *h
	h2.groups = append(h2.groups, name)
	return &h2
}

func (h *NanoHandler) runLoop() {
	defer h.wg.Done()
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	var batch [][]byte

	send := func() {
		if len(batch) == 0 {
			return
		}
		
		// Encode as JSON Array: [ {}, {}, {} ]
		var buf bytes.Buffer
		buf.WriteByte('[')
		for i, b := range batch {
			if i > 0 {
				buf.WriteByte(',')
			}
			buf.Write(b)
		}
		buf.WriteByte(']')

		req, err := http.NewRequest("POST", strings.TrimRight(h.opts.ServerURL, "/")+"/api/ingest/batch", &buf)
		if err == nil {
			req.Header.Set("Content-Type", "application/json")
			req.Header.Set("Authorization", "Bearer "+h.opts.APIKey)
			req.Header.Set("X-Instance-ID", h.instanceID)
			
			client := &http.Client{Timeout: 5 * time.Second}
			resp, err := client.Do(req)
			if err != nil {
				fmt.Fprintf(os.Stderr, "NanoLog Network Error: %v\n", err)
			} else {
				resp.Body.Close()
				if resp.StatusCode != 200 {
					fmt.Fprintf(os.Stderr, "NanoLog Send Failed: HTTP %d\n", resp.StatusCode)
				}
			}
		}

		batch = nil // Reset batch
	}

	for {
		select {
		case data := <-h.queue:
			batch = append(batch, data)
			if len(batch) >= 100 {
				send()
			}
		case <-ticker.C:
			send()
		case <-h.done:
			// Flush remaining
			for {
				select {
				case data := <-h.queue:
					batch = append(batch, data)
				default:
					send()
					return
				}
			}
		}
	}
}

func (h *NanoHandler) Shutdown() {
	close(h.done)
	h.wg.Wait()
}

File: sdks/python/nanolog/handler.py
import logging
import queue
import atexit
import os
import uuid
import json
import socket
import time
import urllib.request
import urllib.error
import sys
from pathlib import Path
from .worker import Worker

class NanoLogHandler(logging.Handler):
    def __init__(self, server_url, api_key, service, source_host=None):
        super().__init__()
        self.server_url = server_url.rstrip('/')
        self.api_key = api_key
        self.service = service
        self.source_host = source_host or socket.gethostname()
        
        self.queue = queue.Queue(maxsize=10000)
        self.instance_id = self._get_or_create_instance_id()
        
        self._register_instance()
        
        self.worker = Worker(self.queue, self.server_url, self.api_key, self.instance_id)
        self.worker.start()
        
        atexit.register(self.shutdown)

    def _get_or_create_instance_id(self):
        home = Path.home()
        nanolog_dir = home / ".nanolog"
        nanolog_dir.mkdir(exist_ok=True)
        id_file = nanolog_dir / "id"
        
        if id_file.exists():
            try:
                return id_file.read_text().strip()
            except Exception:
                pass
        
        new_id = str(uuid.uuid4())
        try:
            id_file.write_text(new_id)
        except Exception:
            pass # Use ephemeral ID if write fails
            
        return new_id

    def _register_instance(self):
        # Handshake with server
        try:
            data = json.dumps({
                "instance_id": self.instance_id,
                "service_name": self.service,
                "host_name": self.source_host,
                "platform": "python",
                "version": "0.1.0"
            }).encode('utf-8')

            req = urllib.request.Request(
                f"{self.server_url}/api/registry/handshake",
                data=data,
                method='POST'
            )
            req.add_header('Content-Type', 'application/json')
            req.add_header('Authorization', f'Bearer {self.api_key}')
            
            with urllib.request.urlopen(req) as response:
                 if response.status != 200:
                    sys.stderr.write(f"NanoLog Handshake Failed: HTTP {response.status}\n")

        except Exception as e:
            sys.stderr.write(f"NanoLog Handshake Error: {e}\n")

    def emit(self, record):
        try:
            # Capture standard logging attributes
            msg = self.format(record)
            
            log_entry = {
                "timestamp": int(record.created * 1000000000), # Nanoseconds
                "level": record.levelname,
                "message": msg,
                "logger": record.name,
                "thread": record.threadName,
                "file": record.filename,
                "line": record.lineno,
                "service": self.service,
                "host": self.source_host,
                "instance_id": self.instance_id,
                "attributes": {}
            }
            
            # Handle extra fields (args) if they are a dict
            if isinstance(record.args, dict):
                 log_entry["attributes"].update(record.args)
                 
            # Trying to extract common trace info if present in record
            if hasattr(record, 'trace_id'):
                 log_entry["trace_id"] = str(record.trace_id)
            if hasattr(record, 'span_id'):
                 log_entry["span_id"] = str(record.span_id)

            try:
                self.queue.put_nowait(log_entry)
            except queue.Full:
                sys.stderr.write("NanoLog Queue Full: Dropping log\n")

        except Exception:
            self.handleError(record)

    def shutdown(self):
        # Stop worker
        if hasattr(self, 'worker') and self.worker.is_alive():
            self.worker.stop()
            self.worker.join()
